### PR TITLE
Add mutual information metric

### PR DESCRIPTION
**Pull request by MooooCat**
## Types of changes

- [ ] Maintenance (no change in code, maintain the project's CI, docs, etc.)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: tests/metrics/test_MISim.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+from sdgx.metrics.pair_column.mi_sim import MISim
+
+# 创建测试数据
+real_data_discrete = pd.DataFrame(
+    {
+        "col1": ["a", "b", "c", "d", "e"],
+        "col2": ["a", "b", "b", "b", "e"],
+    }
+)
```
Comment (MooooCat) [inline tests/metrics/test_MISim.py:13]: please change these test dataset in  `pytest.fixture`(as we discussed before)

```suggestion
@pytest.fixture
def real_data_discrete():
    yield pd.DataFrame(
       {
           "col1": ["a", "b", "c", "d", "e"],
           "col2": ["a", "b", "b", "b", "e"],
       }
   )
```

File: sdgx/metrics/single_table/single_mi_sim.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        """Check the output value.
+
+        Args:
+            raw_metric_value (float):  the calculated raw value of the JSD metric.
+        """
+        # instance = cls()
+        if raw_metric_value < self.lower_bound or raw_metric_value > self.upper_bound:
+            raise ValueError
+
+    # @classmethod
```
Comment (MooooCat) [inline sdgx/metrics/single_table/single_mi_sim.py:73]: These commented codes can be deleted

File: sdgx/metrics/pair_column/mi_sim.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        """Check the output value.
+
+        Args:
+            raw_metric_value (float):  the calculated raw value of the JSD metric.
+        """
+        # instance = cls()
+        if raw_metric_value < self.lower_bound or raw_metric_value > self.upper_bound:
+            raise ValueError
+
+    # @classmethod
```
Comment (MooooCat) [inline sdgx/metrics/pair_column/mi_sim.py:109]: commented codes can be deleted

File: sdgx/metrics/pair_column/mi_sim.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+from sklearn.metrics.cluster import normalized_mutual_info_score
+
+from sdgx.metrics.pair_column.base import PairMetric
+
+
+def Jaccard_index(A, B):
+    return min(A, B) / max(A, B)
+
+
+def time2int(datetime, form):
```
Comment (MooooCat) [inline sdgx/metrics/pair_column/mi_sim.py:16]: can we add this method to our project such as `sdgx.utils.py`?  or integrate this functiona into `formatters`

File: sdgx/metrics/pair_column/mi_sim.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    time_array = time.strptime(datetime, form)
+    time_stamp = int(time.mktime(time_array))
+    return time_stamp
```
Comment (MooooCat) [inline sdgx/metrics/pair_column/mi_sim.py:19]: can we add this method to our project such as `sdgx.utils.py`?  or integrate this functiona into `formatters`

Comment (Z712023) [reply sdgx/metrics/pair_column/mi_sim.py:19]: Good idea

File: sdgx/metrics/pair_column/mi_sim.py (hunk 3 above)
Comment (Z712023) [reply sdgx/metrics/pair_column/mi_sim.py:109]: Well, I need to make a major revision for pair_column/mi_sim.py

File: sdgx/metrics/pair_column/mi_sim.py (hunk 5 above)
Comment (MooooCat) [reply sdgx/metrics/pair_column/mi_sim.py:19]: we may deal with this kind of funcs later when implementing `formatters`

File: sdgx/metrics/multi_table/multitable_mi_sim.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        MISim.check_output(MI_sim)
+
+        return MI_sim
+
+    @classmethod
+    def check_output(raw_metric_value: float):
+        """Check the output value.
+
+        Args:
+            raw_metric_value (float):  the calculated raw value of the JSD metric.
```
Comment (MooooCat) [inline sdgx/metrics/multi_table/multitable_mi_sim.py:67]:
```suggestion
        Args:
            raw_metric_value (float):  the calculated raw value of the mi_sim metric.
```

File: sdgx/metrics/multi_table/multitable_mi_sim.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        self.upper_bound = 1
+        self.metric_name = "mutual_information_similarity"
+        self.numerical_bins = 50
+
+    @classmethod
+    def calculate(
+        real_data: pd.DataFrame, synthetic_data: pd.DataFrame, metadata: dict
+    ) -> pd.DataFrame:
+        """
+        Calculate the JSD value between a real column and a synthetic column.
```
Comment (MooooCat) [inline sdgx/metrics/multi_table/multitable_mi_sim.py:30]:
```suggestion
        Calculate the mi_sim value between a real column and a synthetic column.
```

File: sdgx/metrics/pair_column/base.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        real_data, synthetic_data = PairMetric.check_input(real_data, synthetic_data)
+
+        raise NotImplementedError()
+
+    @classmethod
+    def check_output(cls, raw_metric_value: float):
+        """Check the output value.
+
+        Args:
+            raw_metric_value (float):  the calculated raw value of the JSD metric.
```
Comment (MooooCat) [inline sdgx/metrics/pair_column/base.py:71]:
```suggestion
            raw_metric_value (float):  the calculated raw value of the metric.
```

File: sdgx/metrics/pair_column/mi_sim.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        self.numerical_bins = 50
+
+    @classmethod
+    def calculate(
+        src_col: pd.DataFrame,
+        tar_col: pd.DataFrame,
+        metadata: dict,
+    ) -> float:
+        """
+        Calculate the JSD value between a real column and a synthetic column.
```
Comment (MooooCat) [inline sdgx/metrics/pair_column/mi_sim.py:32]:
```suggestion
        Calculate the metric value between a real column and a synthetic column.
```

File: sdgx/metrics/single_table/single_mi_sim.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        super().__init__()
+        self.lower_bound = 0
+        self.upper_bound = 1
+        self.metric_name = "mutual_information_similarity"
+        self.numerical_bins = 50
+
+    @classmethod
+    def calculate(real_data: pd.DataFrame, synthetic_data: pd.DataFrame, metadata) -> pd.DataFrame:
+        """
+        Calculate the JSD value between a real column and a synthetic column.
```
Comment (MooooCat) [inline sdgx/metrics/single_table/single_mi_sim.py:28]:
```suggestion
        Calculate the mi_sim value between a real column and a synthetic column.
```

File: sdgx/metrics/single_table/single_mi_sim.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        MISim.check_output(MI_sim)
+
+        return MI_sim
+
+    @classmethod
+    def check_output(raw_metric_value: float):
+        """Check the output value.
+
+        Args:
+            raw_metric_value (float):  the calculated raw value of the JSD metric.
```
Comment (MooooCat) [inline sdgx/metrics/single_table/single_mi_sim.py:63]:
```suggestion
            raw_metric_value (float):  the calculated raw value of the mi_sim metric.
```

File: sdgx/utils.py [hunk 12]
```diff
@@ -24,22 +25,9 @@
...
-        "parent_table": "store",
-        "child_table": "train",
-        "parent_url": "https://raw.githubusercontent.com/juniorcl/rossman-store-sales/main/databases/store.csv",
-        "child_url": "https://raw.githubusercontent.com/juniorcl/rossman-store-sales/main/databases/train.csv",
-        "parent_primary_keys": ["Store"],
-        "child_primary_keys": ["Store", "Date"],
-        "foreign_keys": ["Store"],
-    }
-}
-
```
Comment (MooooCat) [inline sdgx/utils.py:42]: These code should not be deleted.

**Review by MooooCat — COMMENTED** [review]
I feel there are still some minor issues that need to be corrected

File: sdgx/utils.py (hunk 12 above)
Comment (Z712023) [reply sdgx/utils.py:42]: Get it
